autoscale/src/main/java/com/sequenceiq/periscope/service/ScalingActivityService.java

```diff
@@ -51,14 +54,25 @@ public ScalingActivity create(Cluster cluster, ActivityStatus activityStatus, St
         return scalingActivityRepository.save(scalingActivity);
     }
 
-    public void updateWithFlowIdAndTriggerStatus(FlowIdentifier flowIdentifier, String activityCrn, ActivityStatus activityStatus) {
-        ScalingActivity scalingActivity = findByCrn(activityCrn);
+    public void update(Long activityId, FlowIdentifier flowIdentifier, ActivityStatus activityStatus, String reason) {
+        ScalingActivity scalingActivity = scalingActivityRepository.findById(activityId).orElseThrow(notFound("ScalingActivity", activityId));
         scalingActivity.setFlowId(flowIdentifier.getPollableId());
         scalingActivity.setActivityStatus(activityStatus);
-        LOGGER.info("Updating ScalingActivity: {} with FlowInformation and ActivityStatus: {}", activityCrn, activityStatus);
+        LOGGER.info("Updating ScalingActivity: {} with FlowInformation and ActivityStatus: {}", scalingActivity.getActivityCrn(), activityStatus);
+        scalingActivity.setScalingActivityReason(reason);
         save(scalingActivity);
     }
 
+    public void setActivityStatusForIds(Collection<Long> ids, ActivityStatus status) {
+        scalingActivityRepository.setActivityStatusesInIds(ids, status);
+    }
+
+    public void setEndTimes(Map<Long, Long> idsWithCompletedFlows) {
```

Comment:
This should be invoked whenever we reach a final state for an activity ? (i.e. the various failure states)

Reply:
Yes, updated to set endTime for all final states